src/workflows/airqo_etl_utils/satellite_data_utils.py

```diff
@@ -0,0 +1,229 @@
+                site_dfs[site] = data_df
+
+            dfs[product] = site_dfs
+
+        return dfs
+
+    def process_site_data(self, dfs):
+
+        site_dfs = {}
+        try:
+            for site in dfs[list(dfs.keys())[0]].keys():
+                products = []
+                # Iterate over each product
+                for product, site_df in dfs.items():
+                    # Retrieve dataframe for the current product and site
+                    product_df = site_df[site]
+                    product_df = product_df.groupby(['date', 'hour']).mean().reset_index()
+                    products.append(product_df)
+
+                # Merge dataframes for all products
+                merged_df = reduce(lambda left, right: pd.merge(left, right, on=['date', 'hour','month'], how='outer'), products)
+                merged_df['site_id'] = site
+                site_dfs[site] = merged_df
+
+        except Exception as e:
+            print(f"Error processing site data: {e}")
+
+        return site_dfs
+    
+
+    def merge_site_data(self, site_dfs):
+        result_df = pd.DataFrame()
+        for key in site_dfs.keys():
+          result_df = pd.concat([result_df, site_dfs[key]])
+        return result_df
+    
+    def extract_and_merge_data(self, df, result_df):
+      df1 = df[['site_id', 'site_name', 'pm2_5', 'pm10', 'date', 'site_latitude', 'site_longitude', 'country', 'month', 'hour']]
+      merged_df_ = df1.merge(result_df, how='right', on=['site_id', 'date', 'month', 'hour'])
+      merged_df_= merged_df_[~merged_df_['country'].isna()].reset_index(drop = True)
+      merged_df_ = merged_df_.drop(columns=[col for col in merged_df_.columns if '_datetime' in col])
+      merged_df_.to_csv('output.csv', index=False)
+
+      return merged_df_
+    
+    def save_to_mongodb(self, merged_df_):
+        try:
+            for record in merged_df_.to_dict(orient='records'):
+                # Check if the record already exists in the database
+                existing_record = self.collection.find_one({
+                    'site_id': record['site_id'],
+                    'date': record['date'],
+                    'hour': record['hour'],
+                    'pm2_5': record['pm2_5'],
+                })
+                if existing_record:
+                    # Update existing record with new data
+                    new_data = {key: value for key, value in record.items() if key not in ['site_id', 'date', 'hour','pm2_5']}
+                    self.collection.update_one(
+                        {'_id': existing_record['_id']},
+                        {'$set': new_data}
+                    )
+                else:
+                    # Insert new record
+                    self.collection.insert_one(record)
+            print("Data saved to MongoDB successfully.")
+        except Exception as e:
+            print(f"Error saving data to MongoDB: {e}")
```

Comment:
Hi @wabinyai mongodb has an `upsert` parameter that allows you to update existing ones, so you can leave out the if else, the code should look somethinig like this after. 
```suggestion
def save_to_mongodb(self, merged_df_):
    try:
        for record in merged_df_.to_dict(orient='records'):
            # Define a filter for identifying the unique document
            filter_ = {
                'site_id': record['site_id'],
                'date': record['date'],
                'hour': record['hour']
            }

            # We use $set to update fields or insert them if they don't exist
            update = {'$set': record}

            # Perform an upsert (update existing or insert new if none exists)
            self.collection.update_one(filter_, update, upsert=True)
    except Exception as e:
        print(f"Error saving data to MongoDB: {e}")

```